master/internal/rm/kubernetesrm/spec.go

```diff
@@ -349,6 +351,43 @@ func (p *pod) createPriorityClass(name string, priority int32) error {
 	return err
 }
 
+const maxChars int = 63
+const fmtAlphaNumeric string = "A-Za-z0-9"
+const fmtAllowedChars string = fmtAlphaNumeric + `\.\-_`
+
+var regDisallowedSpecialChars = regexp.MustCompile("[^" + fmtAllowedChars + "]")
+var regLeadingNonAlphaNumeric = regexp.MustCompile("^[^" + fmtAlphaNumeric + "]+")
+var regTrailingNonAlphaNumeric = regexp.MustCompile("[^" + fmtAlphaNumeric + "]+$")
+
+func validatePodLabelValue(value string) string {
+	errs := validation.IsValidLabelValue(value)
+	if len(errs) == 0 {
+		return value
+	}
+
+	// label value is not valid; attempt to fix it
+	// 0. convert dis-allowed special characters to underscore
+	fixedValue := regDisallowedSpecialChars.ReplaceAllString(value, "_")
+
+	// 1. strip leading non-alphanumeric characters
+	fixedValue = regLeadingNonAlphaNumeric.ReplaceAllString(fixedValue, "")
+
+	// 2. truncate to 63 characters
+	if len(fixedValue) > maxChars {
+		fixedValue = string(fixedValue[:maxChars])
+	}
+
+	// 3. strip ending non-alphanumeric characters
+	fixedValue = regTrailingNonAlphaNumeric.ReplaceAllString(fixedValue, "")
+
+	log.Warnf(
+		"conform to Kubernetes pod label value standards: reformatting %s to %s",
+		value, fixedValue,
+	)
+
+	return fixedValue
```

Comment:
Is it worth a final check that `fixedValue` passes formatting requirements?

Reply:
what would that look like? passing it through validation.IsValidLabelValue(value) again?

Reply:
Yeah. If it fails, we could return an empty string to ensure the workload is still able to be scheduled.

Reply:
I think that's worth a check and you could just return "invalid" as the label if it fails. I think it's OK (maybe ideal) for this code to just return an error though.

Reply:
What would it look like if the code returns an error? Something like,

```
value, err := validatePodLabelValue( v )
if err != nil {
    value = "invalid" // some reasonable default
}
// assign label value
```

Reply:
yeah, i supposed the decision to proceed without the label or not falls to the caller, which makes sense. e.g., i changed reattach in pods to jobs to not rely on the pod [resource pool env var](https://github.com/determined-ai/determined/blob/3b1d0df90302c7ca401ee54390915761a3a06293/master/internal/rm/kubernetesrm/pods.go#L701-L710) but on the resource pool job label on the job. if we can't add it that label, i'll probably need to fail the launch since fault tolerance is lost otherwise.

Reply:
👍 yeah, that makes sense - thanks!